src/Native/LDANative/lda_engine.cpp

```diff
@@ -0,0 +1,1103 @@
+
+        int32_t token_num = 0;
+        int32_t doc_start = data_block_->Begin(thread_id);
+        int32_t doc_end = data_block_->End(thread_id);
+
+        for (int32_t doc_index = doc_start; doc_index != doc_end; ++doc_index)
+        {
+            std::shared_ptr<LDADocument> doc = data_block_->GetOneDoc(doc_index);
+            int doc_size = doc->size();
+            for (int i = 0; i < doc_size; ++i)
+            {
+                int topic = sampler_.rand_k();
+                doc->SetTopic(i, topic);
+            }
+            int cursor = doc->get_cursor();
+            token_num += sampler_.GlobalInit(doc.get());
+        }
+        process_barrier_->wait();
+        CheckFunction(thread_id, tmDebug, "intialize token topic before iterations - in function train_thread");
+
+        for (int i = 0; i < num_threads_; ++i)
+        {
+            std::vector<word_topic_delta>& wtd_vec = samplers_[i]->get_word_topic_delta(thread_id);
+            for (auto& wtd : wtd_vec)
+            {
+                global_word_topic_table_[wtd.word].inc(wtd.topic, wtd.delta);
+            }
+        }
+        process_barrier_->wait();
+        CheckFunction(thread_id, tmDebug, "intialize word topic model before iterations - in function train_thread");
+
+        // use thread-private delta table to get global table
+        {
+            std::lock_guard<std::mutex> lock(atomic_stats_->global_mutex_);
+
+            std::vector<int64_t> &summary = sampler_.get_delta_summary_row();
+            for (int i = 0; i < K_; ++i)
+            {
+                global_summary_row_[i] += summary[i];
+            }
+        }
+        process_barrier_->wait();
+        CheckFunction(thread_id, tmDebug, "global summary & Complete setup train before iterations - in function train_thread");
+
+        for (int iter = 0; iter < num_iterations_; ++iter)
+        {
+            CheckFunction(thread_id, tmDebug, "----------------------iteration start - in function train_thread---------------------");
+            int32_t token_sweeped = 0;
+            atomic_stats_->num_tokens_clock_ = 0;
+            // build alias table
+            // 1, build alias table for the dense term,  beta_k_v_, which is shared by all the words
+            if (thread_id == 0)
+            {
+                beta_mass_ = 0;
+                std::vector<float> proportion(K_);
+                for (int k = 0; k < K_; ++k)
+                {
+                    proportion[k] = beta_ / (global_summary_row_[k] + beta_sum_);
+                    beta_mass_ += proportion[k];
+                }
+
+                alias_rng_int_.SetProportionMass(proportion, beta_mass_, beta_k_v_, &beta_height_, sampler_.rng());
+            }
+            process_barrier_->wait();
+            CheckFunction(thread_id, tmDebug, "built alias table dense - in function train_thread");
+
+            // 2,  build alias table for the sparse term
```

Comment:
>, [](start = 16, length = 1)

so in this code they have
1,
2,
3.
4.
which freaks me out, can we use one thing and I would prefer dot for it?